lib/Dialect/ESI/runtime/cpp/lib/Accelerator.cpp

```diff
@@ -175,25 +175,48 @@ static void loadBackend(std::string backend) {
 namespace registry {
 namespace internal {
 
-static std::map<std::string, BackendCreate> backendRegistry;
-void registerBackend(std::string name, BackendCreate create) {
-  if (backendRegistry.count(name))
-    throw std::runtime_error("Backend already exists in registry");
-  backendRegistry[name] = create;
+class BackendRegistry {
+public:
+  static void registerBackend(const std::string &name, BackendCreate create) {
+    auto &registry = get().backendRegistry;
+    if (registry.count(name))
+      throw std::runtime_error("Backend already exists in registry");
+    registry[name] = create;
+  }
+
+  static std::unique_ptr<AcceleratorConnection>
+  connect(Context &ctxt, const std::string &backend,
+          const std::string &connection) {
+    auto &registry = get().backendRegistry;
+    auto f = registry.find(backend);
+    if (f == registry.end()) {
+      // If it's not already found in the registry, try to load it dynamically.
+      loadBackend(backend);
+      f = registry.find(backend);
+      if (f == registry.end())
+        throw std::runtime_error("Backend '" + backend + "' not found");
+    }
+    return f->second(ctxt, connection);
+  }
+
+private:
+  static BackendRegistry &get() {
+    static BackendRegistry instance;
```

Comment:
Does this not work as a private static pointer? (The usual way one would do this.) I'd hope that C++ would guarantee that all static fields in a particular class are initialized before any calls to static methods (but I may be giving C++ too much credit).